File: app/src/main/java/mx/edu/upqroo/kristenandroid/ui/activities/MainActivity.java
package mx.edu.upqroo.kristenandroid.ui.activities;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;

import com.google.android.material.bottomnavigation.BottomNavigationView;
import com.google.android.material.navigation.NavigationView;
import com.google.android.material.snackbar.Snackbar;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBarDrawerToggle;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.widget.Toolbar;
import androidx.core.view.GravityCompat;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.navigation.NavController;
import androidx.navigation.NavDestination;
import androidx.navigation.Navigation;
import mx.edu.upqroo.kristenandroid.R;
import mx.edu.upqroo.kristenandroid.data.repositories.UserInformationRepository;
import mx.edu.upqroo.kristenandroid.managers.FragmentManager;
import mx.edu.upqroo.kristenandroid.managers.SessionManager;
import mx.edu.upqroo.kristenandroid.services.kristen.KristenApiServices;
import mx.edu.upqroo.kristenandroid.services.kristen.messages.CalendarUrlMessage;
import mx.edu.upqroo.kristenandroid.ui.fragments.CalendarFragment;
import mx.edu.upqroo.kristenandroid.widget.ScheduleWidget;

public class MainActivity extends ThemeActivity
        implements NavigationView.OnNavigationItemSelectedListener,
        BottomNavigationView.OnNavigationItemSelectedListener,
        NavController.OnDestinationChangedListener {
    private Toolbar mToolbar;
    private SessionManager mSession;
    private NavigationView mNavigationView;
    private BottomNavigationView mBottomNavigationView;
    private NavController mNavController;
    private DrawerLayout mDrawerLayout;
    private FragmentManager mFragmentManager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mSession = SessionManager.getInstance();
        onWidgetUpdateMessage(this);
        //region toolbar setup
        setContentView(R.layout.activity_main);
        mToolbar = findViewById(R.id.toolbar);
        mToolbar.setTitle(R.string.nav_menu_news);
        setSupportActionBar(mToolbar);
        //endregion
        //region drawer layout setup
        mDrawerLayout = findViewById(R.id.drawer_layout);
        ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(
                this, mDrawerLayout, mToolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close);
        mDrawerLayout.addDrawerListener(toggle);
        toggle.syncState();

        mNavigationView = findViewById(R.id.nav_view);
        View navHeader = mNavigationView.inflateHeaderView(R.layout.nav_header_news);
        TextView mNavHeaderProfileName = navHeader.findViewById(R.id.text_nav_header_title);
        TextView mNavHeaderProfileEmail = navHeader.findViewById(R.id.text_nav_header_subtitle);
        if (mNavHeaderProfileName != null && mNavHeaderProfileEmail != null) {
            mNavHeaderProfileName.setText(mSession.getSession().getName());
            mNavHeaderProfileEmail.setText(mSession.getSession().getEmail());
        }

        mNavigationView.setNavigationItemSelectedListener(this);

        if (!mSession.sessionAlive()) {
            mNavigationView.getMenu()
                    .getItem(mNavigationView.getMenu().size() - 1)
                    .setTitle(R.string.button_login);
        }
        //endregion
        //region bottom navigation setup
        mBottomNavigationView = findViewById(R.id.bottom_navigation);
        mBottomNavigationView.setOnNavigationItemSelectedListener(this);
        //endregion
        //region nav controller setup
        mNavController = Navigation.findNavController(this, R.id.nav_host_fragment);
        mFragmentManager = FragmentManager.NEWS_FRAGMENT;
        mNavController.addOnDestinationChangedListener(this);
        //endregion
    }

    @Override
    public void onStart() {
        super.onStart();
    }

    @Override
    public void onStop() {
        super.onStop();
        EventBus.getDefault().unregister(this);
    }

    @Override
    public void onBackPressed() {
        if (mDrawerLayout.isDrawerOpen(GravityCompat.START)) {
            mDrawerLayout.closeDrawer(GravityCompat.START);
        } else {
            super.onBackPressed();
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.news, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();

        if (id == R.id.action_settings) {
            startActivity(new Intent(this, SettingsActivity.class));
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean onNavigationItemSelected(@NonNull MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.nav_logout) {
            if (mSession.sessionAlive()) {
                showLogoutDialog();
            } else {
                Intent loginIntent = new Intent(this, LoginActivity.class);
                loginIntent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
                startActivity(loginIntent);
            }
        } else if ((id == R.id.nav_news || id == R.id.news_menu_item)
                && mFragmentManager != FragmentManager.NEWS_FRAGMENT) {

            mNavController.navigate(R.id.newsListFragment);

        } else if (id == R.id.nav_notices
                && mFragmentManager != FragmentManager.NOTICES_FRAGMENT) {

            mNavController.navigate(R.id.noticesFragment);

        } else if ((id == R.id.nav_calendar || id == R.id.calendar_menu_item)
                && mFragmentManager != FragmentManager.CALENDAR_FRAGMENT) {

            if (!EventBus.getDefault().isRegistered(this)) {
                EventBus.getDefault().register(this);
            }
            KristenApiServices.getInstance().getCalendarUrl();

        } else if (id == R.id.nav_settings) {

            mNavController.navigate(R.id.settingsActivity);

        } else {

            if (!mSession.sessionAlive()) {
                Snackbar.make(findViewById(R.id.bottom_navigation),
                        R.string.login_message,
                        Snackbar.LENGTH_LONG)
                        .setAction(R.string.button_login, v -> startActivity(
                                new Intent(getApplicationContext(), LoginActivity.class)))
                        .show();

            } else {

                if (id == R.id.nav_user
                        && mFragmentManager != FragmentManager.USER_FRAGMENT) {

                    mNavController.navigate(R.id.userFragment);

                } else if ((id == R.id.nav_schedule || id == R.id.schedule_menu_item)
                        && mFragmentManager != FragmentManager.SCHEDULE_FRAGMENT) {

                    mNavController.navigate(R.id.scheduleFragment);

                } else if ((id == R.id.nav_grades || id == R.id.grades_menu_item)
                        && mFragmentManager != FragmentManager.GRADES_FRAGMENT) {

                    mNavController.navigate(R.id.gradesFragment);

                } else if (id == R.id.nav_kardex
                        && mFragmentManager != FragmentManager.KARDEX_FRAGMENT) {

                    mNavController.navigate(R.id.kardexFragment);

                }
            }
        }
        mDrawerLayout.closeDrawer(GravityCompat.START);
        return true;
    }

    @Override
    public void onDestinationChanged(@NonNull NavController controller,
                                     @NonNull NavDestination destination,
                                     @Nullable Bundle arguments) {
        mToolbar.setTitle(destination.getLabel());
        switch (destination.getId()) {
            case R.id.newsListFragment:
                mNavigationView.setCheckedItem(R.id.nav_news);
                mBottomNavigationView.getMenu().getItem(0).setChecked(true);
                mBottomNavigationView.setVisibility(View.VISIBLE);

                mFragmentManager = FragmentManager.NEWS_FRAGMENT;
                break;
            case R.id.userFragment:
                mNavigationView.setCheckedItem(R.id.nav_user);
                mBottomNavigationView.setVisibility(View.GONE);

                mFragmentManager = FragmentManager.USER_FRAGMENT;
                break;
            case R.id.scheduleFragment:
                mNavigationView.setCheckedItem(R.id.nav_schedule);
                mBottomNavigationView.getMenu().getItem(1).setChecked(true);
                mBottomNavigationView.setVisibility(View.VISIBLE);

                mFragmentManager = FragmentManager.SCHEDULE_FRAGMENT;
                break;
            case R.id.gradesFragment:
                mNavigationView.setCheckedItem(R.id.nav_grades);
                mBottomNavigationView.getMenu().getItem(2).setChecked(true);
                mBottomNavigationView.setVisibility(View.VISIBLE);

                mFragmentManager = FragmentManager.GRADES_FRAGMENT;
                break;
            case R.id.noticesFragment:
                mNavigationView.setCheckedItem(R.id.nav_notices);
                mBottomNavigationView.setVisibility(View.GONE);

                mFragmentManager = FragmentManager.NOTICES_FRAGMENT;
                break;
            case R.id.kardexFragment:
                mNavigationView.setCheckedItem(R.id.nav_kardex);
                mBottomNavigationView.setVisibility(View.GONE);

                mFragmentManager = FragmentManager.KARDEX_FRAGMENT;
                break;
            case R.id.calendarFragment:
                mNavigationView.setCheckedItem(R.id.nav_calendar);
                mBottomNavigationView.getMenu().getItem(3).setChecked(true);
                mBottomNavigationView.setVisibility(View.VISIBLE);

                mFragmentManager = FragmentManager.CALENDAR_FRAGMENT;
                break;
        }
    }

    protected void onWidgetUpdateMessage(Context context) {
        Intent intent_meeting_update = new Intent(context, ScheduleWidget.class);
        intent_meeting_update.setAction(ScheduleWidget.UPDATE_MEETING_ACTION);
        sendBroadcast(intent_meeting_update);
    }

    private void showLogoutDialog() {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setMessage(getString(R.string.exit_confirmation_message))
                .setCancelable(false)
                .setPositiveButton(getString(R.string.yes_option), (dialog, id) -> {
                    mSession.logout();
                    UserInformationRepository.getInstance(getApplication()).deleteAll();
                    Intent loginIntent = new Intent(getApplicationContext(), LoginActivity.class);
                    loginIntent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK |
                            Intent.FLAG_ACTIVITY_CLEAR_TASK);
                    startActivity(loginIntent);
                })
                .setNegativeButton(getString(R.string.no_option), (dialog, id) -> dialog.cancel());
        AlertDialog alert = builder.create();
        alert.show();
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void calendarServiceResponse(CalendarUrlMessage message) {
        if (EventBus.getDefault().isRegistered(this)) {
            EventBus.getDefault().unregister(this);
        }
        if (!message.getCalendarUrl().isEmpty()) {
            Bundle bundle = new Bundle();
            bundle.putString(CalendarFragment.URL_KEY, message.getCalendarUrl());
            mNavController.navigate(R.id.calendarFragment, bundle);
        }
    }
}

File: app/src/main/java/mx/edu/upqroo/kristenandroid/ui/activities/MainActivity.kt
package mx.edu.upqroo.kristenandroid.ui.activities

import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.widget.TextView
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.widget.Toolbar
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout
import androidx.lifecycle.LiveData
import androidx.lifecycle.Observer
import androidx.navigation.NavController
import androidx.navigation.NavDestination
import androidx.navigation.ui.NavigationUI.setupActionBarWithNavController
import com.google.android.material.appbar.AppBarLayout
import com.google.android.material.bottomnavigation.BottomNavigationView
import com.google.android.material.navigation.NavigationView
import com.google.android.material.snackbar.Snackbar
import mx.edu.upqroo.kristenandroid.R
import mx.edu.upqroo.kristenandroid.api.kristen.KristenApiServices
import mx.edu.upqroo.kristenandroid.data.repositories.UserInformationRepository
import mx.edu.upqroo.kristenandroid.helpers.ScrollToTop
import mx.edu.upqroo.kristenandroid.util.Fragments
import mx.edu.upqroo.kristenandroid.managers.SessionManager
import mx.edu.upqroo.kristenandroid.util.setupWithNavController
import mx.edu.upqroo.kristenandroid.widget.DataWidgetManager
import org.greenrobot.eventbus.EventBus

class MainActivity : ThemeActivity(),
        NavigationView.OnNavigationItemSelectedListener,
        BottomNavigationView.OnNavigationItemSelectedListener,
        NavController.OnDestinationChangedListener {
    private lateinit var mToolbar: Toolbar
    private lateinit var mSession: SessionManager
    private lateinit var mNavigationView: NavigationView
    private lateinit var mBottomNavigationView: BottomNavigationView
    private lateinit var mDrawerLayout: DrawerLayout
    private lateinit var mAppBar: AppBarLayout

    private var mCurrentNavController: LiveData<NavController>? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mSession = SessionManager.instance
        //region widget update
        DataWidgetManager.updateWidgetAsync(applicationContext)
        //endregion
        //region toolbar setup
        setContentView(R.layout.activity_main)
        mToolbar = findViewById(R.id.toolbar)
        mAppBar = findViewById(R.id.main_app_bar)
        setSupportActionBar(mToolbar)
        //endregion
        //region drawer layout setup
        mDrawerLayout = findViewById(R.id.drawer_layout)
        val toggle = ActionBarDrawerToggle(
                this, mDrawerLayout, mToolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close)
        mDrawerLayout.addDrawerListener(toggle)
        toggle.syncState()

        mNavigationView = findViewById(R.id.nav_view)
        val navHeader = mNavigationView.inflateHeaderView(R.layout.nav_header_news)
        val mNavHeaderProfileName = navHeader.findViewById<TextView>(R.id.text_nav_header_title)
        val mNavHeaderProfileEmail = navHeader.findViewById<TextView>(R.id.text_nav_header_subtitle)
        if (mNavHeaderProfileName != null && mNavHeaderProfileEmail != null) {
            mNavHeaderProfileName.text = mSession.session.name
            mNavHeaderProfileEmail.text = mSession.session.email
        }

        mNavigationView.setNavigationItemSelectedListener(this)

        if (!mSession.sessionAlive()) {
            mNavigationView.menu
                    .getItem(mNavigationView.menu.size() - 1)
                    .setTitle(R.string.button_login)
        }
        //endregion
        //region bottom navigation setup
        if (savedInstanceState == null) {
            setupBottomNavigationBar()
        }
        //endregion
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle?) {
        super.onRestoreInstanceState(savedInstanceState)
        setupBottomNavigationBar()
    }

    override fun onBackPressed() {
        if (mDrawerLayout.isDrawerOpen(GravityCompat.START)) {
            mDrawerLayout.closeDrawer(GravityCompat.START)
        } else {
            if (mCurrentNavController?.value?.popBackStack() != true) {
                super.onBackPressed()
            }
        }
    }

    override fun onSupportNavigateUp(): Boolean {
        return mCurrentNavController?.value?.navigateUp() ?: false
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.news_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId

        if (id == R.id.action_settings) {
            startActivity(Intent(this, SettingsActivity::class.java))
        }

        return super.onOptionsItemSelected(item)
    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        when (val id = item.itemId) {
            R.id.nav_logout -> {
                if (mSession.sessionAlive()) {
                    showLogoutDialog()
                    mDrawerLayout.closeDrawer(GravityCompat.START)
                } else {
                    val loginIntent = Intent(this, LoginActivity::class.java)
                    loginIntent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
                    startActivity(loginIntent)
                    mDrawerLayout.closeDrawer(GravityCompat.START)
                }
            }
            R.id.nav_settings -> {
                startActivity(Intent(this, SettingsActivity::class.java))
                mDrawerLayout.closeDrawer(GravityCompat.START)
            }
            R.id.nav_contact -> {
                openDetailActivity(Fragments.CONTACT_FRAGMENT.value)
                mDrawerLayout.closeDrawer(GravityCompat.START)
            }
            else -> {
                if (!mSession.sessionAlive()) {
                    showLoginMessage()
                } else {
                    when (id) {
                        R.id.nav_user -> openDetailActivity(Fragments.USER_FRAGMENT.value)
                        R.id.nav_kardex -> openDetailActivity(Fragments.KARDEX_FRAGMENT.value)
                        R.id.nav_notices -> openDetailActivity(Fragments.NOTICES_FRAGMENT.value)
                    }
                    mDrawerLayout.closeDrawer(GravityCompat.START)
                }
            }
        }
        return false
    }

    override fun onDestinationChanged(controller: NavController,
                                      destination: NavDestination,
                                      arguments: Bundle?) {
        val id = destination.id
        mAppBar.setExpanded(true, true)
        if (id == R.id.calendarFragment) {
            KristenApiServices.instance.getCalendarUrl()
        }
    }

    private fun setupBottomNavigationBar() {
        mBottomNavigationView = findViewById(R.id.bottom_navigation)
        val navGraphIds = listOf(R.navigation.news, R.navigation.schedule, R.navigation.calendar,
                R.navigation.grades)

        // Setup the bottom navigation view with a list of navigation graphs
        val controller = mBottomNavigationView.setupWithNavController(
                navGraphIds = navGraphIds,
                fragmentManager = supportFragmentManager,
                containerId = R.id.nav_host_fragment,
                intent = intent
        )

        mBottomNavigationView.setOnNavigationItemReselectedListener {
            mAppBar.setExpanded(true, true)
            EventBus.getDefault().post(ScrollToTop(it.itemId))
        }

        // Whenever the selected controller changes, setup the action bar.
        controller.observe(this, Observer { navController ->
            setupActionBarWithNavController(this, navController)
            navController.addOnDestinationChangedListener(this)
        })
        mCurrentNavController = controller
    }

    private fun openDetailActivity(fragmentToShow: String) {
        val intent = Intent(this, DetailsActivity::class.java)
        intent.putExtra(DetailsActivity.FRAGMENT_TO_SHOW, fragmentToShow)
        startActivity(intent)
    }

    private fun showLoginMessage() {
        Snackbar.make(findViewById(R.id.bottom_navigation),
                R.string.login_message,
                Snackbar.LENGTH_LONG)
                .setAction(R.string.button_login) { startActivity(
                        Intent(applicationContext, LoginActivity::class.java)) }
                .show()
    }

    private fun showLogoutDialog() {
        val builder = AlertDialog.Builder(this)
        builder.setMessage(getString(R.string.exit_confirmation_message))
                .setCancelable(false)
                .setPositiveButton(getString(R.string.yes_option)) { _, _ ->
                    mSession.logout()
                    DataWidgetManager.deleteSchedule(applicationContext)
                    UserInformationRepository.getInstance(application).deleteAll()
                    val loginIntent = Intent(applicationContext, LoginActivity::class.java)
                    loginIntent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
                    startActivity(loginIntent)
                }
                .setNegativeButton(getString(R.string.no_option)) { dialog, _ -> dialog.cancel() }
        val alert = builder.create()
        alert.setOnShowListener {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                alert.getButton(AlertDialog.BUTTON_NEGATIVE).setTextColor(getColor(R.color.colorAccent))
                alert.getButton(AlertDialog.BUTTON_POSITIVE).setTextColor(getColor(R.color.colorAccent))
            }
        }
        alert.show()
    }
}
